### PR TITLE
KAFKA-5378: Return LSO in FetchResponse plus some metrics

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @ijuma @junrao Let me know if you think the metric changes are reasonable.

**Comment by junrao** [comment]
@hachikuji : Thanks for the patch. LGTM. I will let you merge it.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 1]
```diff
@@ -124,28 +124,28 @@ public void setup() {
 
     @Test(expected = IllegalStateException.class)
     public void testFailIfUnreadyForSendNoProducerId() {
-        transactionManager.failIfUnreadyForSend();
+        transactionManager.failIfNotReadyForSend();
     }
 
     @Test
     public void testFailIfUnreadyForSendIdempotentProducer() {
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:131]: I changed a few test names to match the rename for the methods they are testing.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 2]
```diff
@@ -1292,30 +1292,6 @@ public void testAllowDrainInAbortableErrorState() throws InterruptedException {
         assertTrue(transactionManager.hasAbortableError());
     }
 
-    @Test
-    public void testRaiseErrorWhenNoPartitionsPendingOnDrain() throws InterruptedException {
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:1296]: I guess this test is no longer relevant?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 3]
```diff
@@ -684,11 +690,15 @@ class ReplicaManager(val config: KafkaConfig,
...
+        val initialHighWatermark = localReplica.highWatermark.messageOffset
+        val lastStableOffset = if (isolationLevel == IsolationLevel.READ_COMMITTED)
           Some(localReplica.lastStableOffset.messageOffset)
-        else if (readOnlyCommitted)
-          Some(localReplica.highWatermark.messageOffset)
+        else
+          None
+
+        // decide whether to only fetch committed data (i.e. messages below high watermark)
+        val maxOffsetOpt = if (readOnlyCommitted)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:700]: It's a bit confusing that we now have 2 slightly different meanings for `committed`.

**Review by ijuma — APPROVED** [review]
Thanks, LGTM. I fixed some minor things before merging as per my comments below. Also, I didn't review the changes in `ReplicaManagerTest` in detail, relying on Jun for that. :) Merging to trunk and 0.11.0.

**Pull request closed without merging**